src/IB/IBFEMethod.cpp

```diff
@@ -224,15 +229,427 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
     return;
 }
 
-static const Real PENALTY = 1.e10;
+// initial condition for the stress normalization function Phi
+// which solves the heat equation.
+Number
+heat_initial_condition(const libMesh::Point& p, const Parameters& parameters, const std::string&, const std::string&)
+{
+    const Real x = p(0);
+    const Real y = p(1);
+    const Real z = p(2);
+
+    return 0.0;
+}
+
+void
+assemble_cg_heat(EquationSystems& es, const std::string& system_name)
+{
+    const MeshBase& mesh = es.get_mesh();
+    const BoundaryInfo& boundary_info = *mesh.boundary_info;
+    const unsigned int dim = mesh.mesh_dimension();
+    TransientLinearImplicitSystem& system = es.get_system<TransientLinearImplicitSystem>(IBFEMethod::PHI_SYSTEM_NAME);
+    const DofMap& dof_map = system.get_dof_map();
+    FEType fe_type = dof_map.variable_type(0);
+
+    UniquePtr<FEBase> fe(FEBase::build(dim, fe_type));
+    QGauss qrule(dim, FIFTH);
+    fe->attach_quadrature_rule(&qrule);
+    UniquePtr<FEBase> fe_face(FEBase::build(dim, fe_type));
+    QGauss qface(dim - 1, FIFTH);
+    fe_face->attach_quadrature_rule(&qface);
+
+    const std::vector<Real>& JxW = fe->get_JxW();
+    const std::vector<std::vector<Real> >& phi = fe->get_phi();
+    const std::vector<std::vector<RealGradient> >& dphi = fe->get_dphi();
+
+    const std::vector<std::vector<Real> >& phi_face = fe_face->get_phi();
+    const std::vector<Real>& JxW_face = fe_face->get_JxW();
+
+    DenseMatrix<Number> Ke;
+    std::vector<dof_id_type> dof_indices;
+
+    const Real dt = es.parameters.get<Real>("dt");
```

Comment:
Need to get the time step size dynamically, or generate an error message if the actual time step size is different from the expected one. In general, we try to make the solvers work "in principle" with variable time step sizes, so that if we ever figure out how to estimate the time step size, we can do it; or so that if we ever get a really efficient implicit solver, it too can dynamically choose its time step sizes.

Reply:
would there some way to tell when the timestep changes, or should I just track two consecutive timestep sizes and see if they are the same?

Reply:
in other words, i want to assemble the matrix only when the timestep changes...

Reply:
@boyceg what do you think?

Reply:
If you can access the member functions for the `HierarchyIntegrator` in this assembly function, then it should be easy to detect.

However, the better thing to do is to make this function always rebuild the full system, but then only call the full assembly function when needed. You can set up a `System` not to reassemble the full system matrix when you call `System::solve()` by setting the member variable `System::assemble_before_solve` to `false`. Then you would just assemble as needed.

You would want to split out the matrix and RHS assembly routines to make this work.

Reply:
this is how I am doing things now I think.   computeStressNormalization populates the rhs vector, and then I set assemble_before_solve to false.  I just need to fix how I am handling the timestep then?

Reply:
Yes I think so.

I am not sure the best way to communicate the time step size to the solver. I am not crazy about putting extra parameters like the time step size in as a parameter for the full `EquationSystems` object, but that seems to be the usual libMesh way to get parameters into these kinds of routines.

What I do in `FEDataManager` is not to use the libMesh assembly infrastructure, but just to do it myself.

An alternative would be to provide a `libMesh::Assembly` object (see [here](http://libmesh.github.io/doxygen/classlibMesh_1_1System_1_1Assembly.html)) to do the assembly and store the system-specific parameters.

Reply:
If you use `libMesh::Assembly`, you just would have a little class that has an `assemble` member function instead of using a static function to assemble, as in the current implementation.

If you did this, then you could put all of the CG and IPDG parameters directly in the relevant `Assembly` classes. This would cut down a bit on declaring the same variable names over and over. :-)